File: 89. Gray Code.java
// 89. Gray Code

class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> code = new ArrayList<>();
        
        for (int i = 0; i < 1 << n; i++) {
            code.add(i ^ (i >> 1));
        }

        return code;
    }
}

// class Solution {
//     public List<Integer> greyCode(int n) {
//         List<Integer> code = new ArrayList<>();
//         code.add(0);
//         for(int mask=0; mask<n; mask++) {
//             for(int prev=code.size()-1; prev>=0; prev--) {
//                 code.add((1 << mask) | code.get(prev));
//             }
//         }
//         return code;
//     }
// }

File: 89. Gray Code.cpp
class Solution {
public:
    vector<int> grayCode(int n) {
        vector<int> res;
        res.push_back(0)
        if(n == 0) return res;
        for(int i=1; i<n; i++) {
            int add= 1<<i;
            int end = res.size()-1;
            while(end >= 0) {
                res.push_back(res[end--] + add);
            }
        }
        return res;
    }
};
